File: frontend/src/lib/services/favorite.service.ts
import { apiClient } from '@/lib/api';
import type { Favorite, ApiResponse } from '@/types';

export const favoriteService = {
    async addFavorite(teacherId: number): Promise<ApiResponse> {
        return apiClient.post('/favorites', { teacher_id: teacherId });
    },

    async removeFavorite(teacherId: number): Promise<ApiResponse> {
        return apiClient.delete(`/favorites/${teacherId}`);
    },

    async getFavorites(limit = 10, offset = 0): Promise<ApiResponse<{ favorites: Favorite[]; total: number }>> {
        return apiClient.get(`/favorites?limit=${limit}&offset=${offset}`);
    },

    async checkFavorite(teacherId: number): Promise<ApiResponse<{ is_favorite: boolean }>> {
        return apiClient.get(`/favorites/check/${teacherId}`);
    },
};

File: backend/internal/service/favorite_service.go
package service

import (
	"github.com/timebankingskill/backend/internal/models"
	"github.com/timebankingskill/backend/internal/repository"
)

type FavoriteService interface {
	AddFavorite(userID, teacherID uint) error
	RemoveFavorite(userID, teacherID uint) error
	GetFavorites(userID uint, limit, offset int) ([]models.Favorite, int64, error)
	IsFavorite(userID, teacherID uint) (bool, error)
}

type favoriteService struct {
	repo repository.FavoriteRepository
}

func NewFavoriteService(repo repository.FavoriteRepository) FavoriteService {
	return &favoriteService{repo: repo}
}

func (s *favoriteService) AddFavorite(userID, teacherID uint) error {
	exists, err := s.repo.Exists(userID, teacherID)
	if err != nil {
		return err
	}
	if exists {
		return nil // Already favorited
	}

	favorite := &models.Favorite{
		UserID:    userID,
		TeacherID: teacherID,
	}
	return s.repo.Create(favorite)
}

func (s *favoriteService) RemoveFavorite(userID, teacherID uint) error {
	return s.repo.Delete(userID, teacherID)
}

func (s *favoriteService) GetFavorites(userID uint, limit, offset int) ([]models.Favorite, int64, error) {
	return s.repo.GetByUserID(userID, limit, offset)
}

func (s *favoriteService) IsFavorite(userID, teacherID uint) (bool, error) {
	return s.repo.Exists(userID, teacherID)
}
